ASoC/SoundWire: handle cpu_dais before codec_dais

plbossart opened the pull request:
This PR corrects a weird order enforced by ASoC/DPCM where specifically hw_params() will be handled for codec-dais before cpu-dais. This is particularly odd in the case of SoundWire, where we end-up allocating master runtime objects from the codec driver - most likely as a workaround.

plbossart commented:
https://sof-ci.01.org/linuxpr/PR3125/build6250/devicetest/ shows two alsa-bat failures on HDaudio platforms, re-running the tests.

plbossart commented:
SOFCI TEST

plbossart commented:
alsabat problems confirmed by https://sof-ci.01.org/linuxpr/PR3125/build6251/devicetest/

@ranj063 any idea what could be the problem with HDA DAIs?

plbossart commented:
Tested on HDaudio platform,  no sound generated with this PR.

Reverting "ASoC: soc-pcm: enforce order between cpu and codec dais" make the speakers/headset audible again. Wondering if this is a hidden reason behind the specific trigger order for HDAudio...

ranj063 commented:
> alsabat problems confirmed by https://sof-ci.01.org/linuxpr/PR3125/build6251/devicetest/
> 
> @ranj063 any idea what could be the problem with HDA DAIs?

@plbossart I wonder if the "snd_soc_dapm_update_dai(substream, params, cpu_dai);" while doing the hw_params for the cpu_dai first is the reason for the failure

plbossart commented:
@lgirdwood it seems that since the initial commit ddee627cf6bb601aa980104fc17d4f84201380be ('ASoC: core - Separate out PCM operations into new file.') the hw_params for the codec_dai was done *before* the hw_params for the cpu_dai.

Any specific reason why that was the case?

There were simplifications in 93e6958a3674d2fa42e2c24ad5156e65da1d8621 and later, but that order has not changed since the beginning. Was this by design?

plbossart commented:
> > alsabat problems confirmed by https://sof-ci.01.org/linuxpr/PR3125/build6251/devicetest/
> > @ranj063 any idea what could be the problem with HDA DAIs?
> 
> @plbossart I wonder if the "snd_soc_dapm_update_dai(substream, params, cpu_dai);" while doing the hw_params for the cpu_dai first is the reason for the failure

no, I tried with the updated patches. It's really the cpu_dai hw_params that needs to be done last. 

@bardliao @kv2019i @ranj063 @RanderWang any idea why this is needed for HDaudio?

plbossart commented:
My brain is starting to explode. The code in hda-dai.c seems to mix dai and dailink bigtime. In a dai ops for the cpu-dai, we modify the codec dai.

````c
        /* set the stream tag in the codec dai dma params */
	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		snd_soc_dai_set_tdm_slot(codec_dai, stream_tag, 0, 0, 0);
	else
		snd_soc_dai_set_tdm_slot(codec_dai, 0, stream_tag, 0, 0);

      if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		p_params.link_bps = codec_dai->driver->playback.sig_bits;
	else
		p_params.link_bps = codec_dai->driver->capture.sig_bits;

	return hda_link_dma_params(link_dev, &p_params);
}
````

clearly something wrong here in terms of layering violations.  the cpu dai should not deal with codec dai information...

bardliao commented:
@plbossart @ranj063 I tested "HDA Analog" playback on my Up board and the sound seems good to me. I also ran check-alsabat.sh on CML-SKU0955-HDA and it passed. So I don't know how it failed on the devicetest.

plbossart commented:
> @plbossart @ranj063 I tested "HDA Analog" playback on my Up board and the sound seems good to me. I also ran check-alsabat.sh on CML-SKU0955-HDA and it passed. So I don't know how it failed on the devicetest.

interesting, so that was a test on cAVS 1.8. I tested on TGL, i'll try again on the UpExtreme.

plbossart commented:
@bardliao I am able to reproduce your test, this PR does work on UpExtreme (CNL). For some reason it doesn't on my TGL device. I rebased and will re-check test results but as things stand this looks like a platform dependency we never captured.

@ranj063 @kv2019i would you happen to have any TGL / HDAudio devices that you can try this PR on? Thanks!

ranj063 commented:
> @ranj063 @kv2019i would you happen to have any TGL / HDAudio devices that you can try this PR on? Thanks!

@plbossart I have to TGL/HDAudio devices but wouldnt the HDMI playback on TGL I2S devices also be affected by this PR?

plbossart commented:
> > @ranj063 @kv2019i would you happen to have any TGL / HDAudio devices that you can try this PR on? Thanks!
> 
> @plbossart I have to TGL/HDAudio devices but wouldnt the HDMI playback on TGL I2S devices also be affected by this PR?

It would, it's just less straightforward to test when compounding DRM/HDMI issues.

plbossart commented:
I managed to root-cause the issue with HDaudio dais, it's related to our use of the set_tdm_mode 
````c
        /* set the stream tag in the codec dai dma params */
	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		snd_soc_dai_set_tdm_slot(codec_dai, stream_tag, 0, 0, 0);
	else
		snd_soc_dai_set_tdm_slot(codec_dai, 0, stream_tag, 0, 0);
````

This is set in the codec_dai, but with the current order the hw_params of the codec_dai is executed before this. When the order is reversed, then the tx_mask is used in the codec_dai fixup and understood as a channel mask - when it's really a linear value.

The only 'clean' way to solve this is to stop overloading the tx_mask and use a cleaner get/set_stream API, building on what already exists for SoundWire.

@bardliao @RanderWang @ranj063 @kv2019i @ujfalusi comments welcome. This is a massively invasive patch so let's be careful.

bardliao commented:
@plbossart The change looks good to me. Do we need to add some comments in soc-pcm.c so people will not try to revise the order in the future?

kv2019i commented on sound/soc/intel/skylake/skl-pcm.c:
```diff
@@ -560,19 +560,16 @@ static int skl_link_hw_params(struct snd_pcm_substream *substream,
-	else
-		snd_soc_dai_set_tdm_slot(codec_dai, 0, stream_tag, 0, 0);
+	/* set the stream tag in the codec dai */
+	snd_soc_dai_set_stream(codec_dai, p_stream, substream->stream);
```
Could be simpler to just have:
struct hdac_stream *hstream = hdac_stream(link_dev);
snd_soc_dai_set_stream(codec_dai, &hstream->stream_tag, substream->stream);

And:
p_params.link_dma_id = hstream->stream_tag  - 1

.. no need to refer to hdaudio.h separately and compiler will check the types.

plbossart replied:
Good suggestion. Maybe I should just pass the hdac_stream structure directly, not just the stream tag.

plbossart replied:
done

ujfalusi commented:
@plbossart, is there anything preventing us in case of HDA to use the `snd_soc_dai_set_dma_data()` / `snd_soc_dai_get_dma_data()` to communicate the `stream_id` (which is kind of _the_ DMA data in HDA)?
Use of tx_mask for two unrelated configuration is indeed just asking for trouble to happen.

kv2019i reviewed (commented):
An alternative way to implement the hdac_hda changes inline.

Otherwise looks like a welcome change, although I wonder will other platforms blow up when the order is changed. But yeah, that's a separate discussion to be had on alsa-devel.

ujfalusi commented:
I think the ordering change should be safe for most if not all platforms. SDW and HDA are the ones which have close integration of CPU and CODEC side. For others it does not really matter who goes first with hw_params, CPUs don't tend to care/know about codecs and vice versa.

The `prepare` just goes through all the dais in the link, so no ordering is enforced there.

The hw_params sequence change looks reasonable.

RanderWang commented:
> I managed to root-cause the issue with HDaudio dais, it's related to our use of the set_tdm_mode
> 
> ```c
>         /* set the stream tag in the codec dai dma params */
> 	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
> 		snd_soc_dai_set_tdm_slot(codec_dai, stream_tag, 0, 0, 0);
> 	else
> 		snd_soc_dai_set_tdm_slot(codec_dai, 0, stream_tag, 0, 0);
> ```
> 
> This is set in the codec_dai, but with the current order the hw_params of the codec_dai is executed before this. When the order is reversed, then the tx_mask is used in the codec_dai fixup and understood as a channel mask - when it's really a linear value.
> 
> The only 'clean' way to solve this is to stop overloading the tx_mask and use a cleaner get/set_stream API, building on what already exists for SoundWire.
> 
> @bardliao @RanderWang @ranj063 @kv2019i @ujfalusi comments welcome. This is a massively invasive patch so let's be careful.

I really recall the tx_mask overloading issue two years ago. It is a good idea to go with get/set_stream. I don't see any risk.

plbossart commented:
> @plbossart, is there anything preventing us in case of HDA to use the `snd_soc_dai_set_dma_data()` / `snd_soc_dai_get_dma_data()` to communicate the `stream_id` (which is kind of _the_ DMA data in HDA)?
> Use of tx_mask for two unrelated configuration is indeed just asking for trouble to happen.

Fair question @ujfalusi: we do use the set_dma_data for the HDaudio CPU DAI

````
	/* get stored dma data if resuming from system suspend */
	link_dev = snd_soc_dai_get_dma_data(dai, substream);
````

the point is that the *codec_dai* needs to know about the stream_tag as well. I believe we could probably provide the stream_tag with the set_dma_data for HDaudio, but in the general case set/get_dma_data() is more of a 'private' capability. 
For example in the case of SoundWire, we use dai_get_dma_data to keep a context and a set of flags to deal with suspend, etc. We would not want anyone to override this structure.

The intent of the 'set_stream' API was to avoid interfering from a different layer with private handling at the DAI level of set/get_dma_data()

makes sense? I can add this explanation on why this extension to use set_stream makes sense in general and is not a replacement for get/get_dma_data.

plbossart commented:
> Otherwise looks like a welcome change, although I wonder will other platforms blow up when the order is changed. But yeah, that's a separate discussion to be had on alsa-devel.

I was thinking about this and maybe the 'safe' way to make progress is to add yet another flag at the dailink level to specify the order. default would be zero - codec_dai first, then cpu_dai first - and for SoundWire we tag the dailinks as cpu-dai-first. HDaudio can remain as default zero for now, but at least it can work either way.

The benefit of this flag is that we would not break any legacy where this codec then cpu order is required because of history.

ujfalusi commented:
> > @plbossart, is there anything preventing us in case of HDA to use the `snd_soc_dai_set_dma_data()` / `snd_soc_dai_get_dma_data()` to communicate the `stream_id` (which is kind of _the_ DMA data in HDA)?
> > Use of tx_mask for two unrelated configuration is indeed just asking for trouble to happen.
> 
> Fair question @ujfalusi: we do use the set_dma_data for the HDaudio CPU DAI
> 
> ```
> 	/* get stored dma data if resuming from system suspend */
> 	link_dev = snd_soc_dai_get_dma_data(dai, substream);
> ```
> 
> the point is that the _codec_dai_ needs to know about the stream_tag as well. I believe we could probably provide the stream_tag with the set_dma_data for HDaudio, but in the general case set/get_dma_data() is more of a 'private' capability.
> For example in the case of SoundWire, we use dai_get_dma_data to keep a context and a set of flags to deal with suspend, etc. We would not want anyone to override this structure.
> 
> The intent of the 'set_stream' API was to avoid interfering from a different layer with private handling at the DAI level of set/get_dma_data()
> 
> makes sense? I can add this explanation on why this extension to use set_stream makes sense in general and is not a replacement for get/get_dma_data.

So, the stream_tag is at our fingertips: `hdac_stream(link_dev)->stream_tag` via the link_dev?
We do a snd_soc_dai_get_dma_data() in the pcm callbacks.

plbossart commented:
> So, the stream_tag is at our fingertips: `hdac_stream(link_dev)->stream_tag` via the link_dev?
> We do a snd_soc_dai_get_dma_data() in the pcm callbacks.

not in the codec_dai, see sound/soc/codecs/hdac_hda.c. We already use set_tdm_slots() to pass the stream tag.

plbossart commented:
I updated this PR to make the order opt-in so that we don't break.
I also removed the SoundWire change  in the stream API. it turns out this was counter productive, we want to move all the allocation that takes place during the hw_params() stage into the startup() stage, and there is no specific order enforce there, so the stream API has to deal with any order for allocation. I still think we need to handled the CPU dais first in hw_params due to the completely unbalanced relationship between host and device, all the hard work is done on the cpu side and if that fails there's no point in doing the hw_params on the codec side first.

Still not ready to be merged, I need more time when I am not in meetings to complete this work.

ujfalusi commented:
> > So, the stream_tag is at our fingertips: `hdac_stream(link_dev)->stream_tag` via the link_dev?
> > We do a snd_soc_dai_get_dma_data() in the pcm callbacks.
> 
> not in the codec_dai, see sound/soc/codecs/hdac_hda.c. We already use set_tdm_slots() to pass the stream tag.

Ah, I see. What confused me (and continues to do so) is the fact that the cpu_dai directly talks and configures the codec_dai (it could be the other way around as well, not checked).
The `snd_soc_dai_set_tdm_slot` was definitely abused, it should be gone, but that one should have been called from the machine driver level (dai_link).

For me a `hdac_hda_dai_set_stream_tag(struct snd_soc_dai *dai, u8 stream_tag );` would have been similarly acceptable given the tight coupling of cpu and codec dai of HDA.

As for the new `.set_stream` and `snd_soc_dai_set_stream()`: both SDW and HDA vocabulary includes heavy use of 'stream', I can not even follow how many different **stream** we have in HDA related codes (snd_pcm_substream, hda_pcm_stream, hdac_stream, hdac_ext_stream, sof_intel_hda_stream to name a few), but I have hard time to imagine what **stream** the `snd_soc_dai_set_stream()` sets.

Edit: I know that we are giving the `stream_tag` from `struct hdac_stream`, which is part of `struct hdac_ext_stream`, which is the dma_data of the cpu_dai.

plbossart commented:
> Ah, I see. What confused me (and continues to do so) is the fact that the cpu_dai directly talks and configures the codec_dai (it could be the other way around as well, not checked).
> The `snd_soc_dai_set_tdm_slot` was definitely abused, it should be gone, but that one should have been called from the machine driver level (dai_link).

Right, the problem is is how to introduce dailink ops. We have a huge legacy of existing machine drivers, it's not super appealing to modify all the code in sound/soc/intel/boards/. In addition the code that we have has SOF-specific code, which is problematic since some machine drivers are shared with the SKL driver. I really don't know how to solve this.
> 
> For me a `hdac_hda_dai_set_stream_tag(struct snd_soc_dai *dai, u8 stream_tag );` would have been similarly acceptable given the tight coupling of cpu and codec dai of HDA.
> 
> As for the new `.set_stream` and `snd_soc_dai_set_stream()`: both SDW and HDA vocabulary includes heavy use of 'stream', I can not even follow how many different **stream** we have in HDA related codes (snd_pcm_substream, hda_pcm_stream, hdac_stream, hdac_ext_stream, sof_intel_hda_stream to name a few), but I have hard time to imagine what **stream** the `snd_soc_dai_set_stream()` sets.
> 
> Edit: I know that we are giving the `stream_tag` from `struct hdac_stream`, which is part of `struct hdac_ext_stream`, which is the dma_data of the cpu_dai.

agree, the 'stream' for ALSA means 'direction' or more abstract context. Then we have hstream and sof_stream. it's horrible indeed.

plbossart commented:
closing for now, will submit a new PR when we have fewer in-flight changes. It's not productive at the moment to ask for review/tests.
